src/builder.rs

```diff
@@ -0,0 +1,54 @@
+use crate::{errors::Error, Layer};
+
+#[derive(Debug)]
+pub struct PacketBuilder {
+    layers: Vec<Box<dyn Layer + Send>>,
+    raw_bytes: Option<Vec<u8>>,
+}
+
+impl PacketBuilder {
```

Comment:
I have a very high level comment about the `PacketBuilder` struct and it's APIs. Intuitively one would expect `PacketBuilder` to build the `Packet` structure. However this structure is very different from the original structure and there is considerable divergence. The challenge with that is - for the user it's two sets of APIs (and `struct`s) one for building packets and one for dissecting packets - that's a bit un-intuitive. 

`PacketBuilder::build` should eventually generate a `Packet` structure.  (Note: it strictly doesn't have to be so, but then it should not be called a `PacketBuilder`.)

Reply:
Another way to think about `PacketBuilder` is - it should look something like - 

```rust 
pub struct PacketBuilder {
    inner: Packet,
    // Some more metadata, not sure what that is
}
```

And the `layers`, `raw_bytes` can then be used from the `inner` `Packet` structure. There will be more metadata to the `Packet` structure. This will be used for things like `checksum` and so on.

Reply:
Makes sense, seems like I overlooked the Packet structure.